File: pipelines/lib/registry-loader.js
import fs from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'

const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)
const ROOT_DIR = path.resolve(__dirname, '..', '..')
const REGISTRY_PATH = path.join(ROOT_DIR, 'data', 'registry', 'DATA_REGISTRY.yaml')

let cachedRegistry = null

function minimalParseRegistry(content) {
  const datasets = {}
  let inDatasets = false
  let currentId = null

  for (const rawLine of content.split(/\r?\n/)) {
    const line = rawLine.replace(/\s+$/, '')
    if (!line || line.trimStart().startsWith('#')) continue

    if (line === 'datasets:') {
      inDatasets = true
      continue
    }

    if (!inDatasets) continue

    if (line.startsWith('  ') && !line.startsWith('    ') && line.endsWith(':')) {
      currentId = line.trim().slice(0, -1)
      datasets[currentId] = {}
      continue
    }

    if (currentId && line.startsWith('    ')) {
      const trimmed = line.trim()
      const separatorIndex = trimmed.indexOf(':')
      if (separatorIndex === -1) continue
      const key = trimmed.slice(0, separatorIndex)
      const value = trimmed.slice(separatorIndex + 1).trim().replace(/^['"]|['"]$/g, '')
      datasets[currentId][key] = value
    }
  }

  return { datasets }
}

export function loadRegistry() {
  if (cachedRegistry) return cachedRegistry
  if (!fs.existsSync(REGISTRY_PATH)) {
    cachedRegistry = { datasets: {} }
    return cachedRegistry
  }

  const content = fs.readFileSync(REGISTRY_PATH, 'utf8')
  cachedRegistry = minimalParseRegistry(content)
  return cachedRegistry
}

export function getDatasetPath(datasetId, fallback = null) {
  const datasets = loadRegistry().datasets ?? {}
  const pathValue = datasets?.[datasetId]?.path
  if (pathValue) return path.resolve(ROOT_DIR, pathValue)
  if (fallback == null) {
    throw new Error(`Dataset '${datasetId}' not found in ${REGISTRY_PATH}`)
  }
  return path.isAbsolute(fallback) ? fallback : path.resolve(ROOT_DIR, fallback)
}

File: pipelines/support-data/lib/registry_loader.py
#!/usr/bin/env python3
"""Helpers for loading dataset paths from data/registry/DATA_REGISTRY.yaml."""

from __future__ import annotations

from functools import lru_cache
from pathlib import Path
from typing import Any

try:
    import yaml
except ImportError:  # pragma: no cover - optional dependency
    yaml = None


ROOT_DIR = Path(__file__).resolve().parents[3]
REGISTRY_PATH = ROOT_DIR / "data" / "registry" / "DATA_REGISTRY.yaml"


def _minimal_parse_registry(content: str) -> dict[str, Any]:
    datasets: dict[str, dict[str, Any]] = {}
    in_datasets = False
    current_id: str | None = None
    current_list_key: str | None = None

    for raw_line in content.splitlines():
        line = raw_line.rstrip()
        if not line or line.lstrip().startswith("#"):
            continue

        if line == "datasets:":
            in_datasets = True
            continue

        if not in_datasets:
            continue

        if line.startswith("  ") and not line.startswith("    ") and line.endswith(":"):
            current_id = line.strip()[:-1]
            datasets[current_id] = {}
            current_list_key = None
            continue

        if current_id and line.startswith("    "):
            stripped = line.strip()
            if current_list_key and stripped.startswith("- "):
                datasets[current_id].setdefault(current_list_key, []).append(stripped[2:].strip().strip("'\""))
                continue

            key, sep, value = stripped.partition(":")
            if sep:
                cleaned_value = value.strip().strip("'\"")
                if cleaned_value:
                    datasets[current_id][key] = cleaned_value
                    current_list_key = None
                else:
                    datasets[current_id][key] = []
                    current_list_key = key

    return {"datasets": datasets}


@lru_cache(maxsize=1)
def load_registry() -> dict[str, Any]:
    """Load and cache the registry payload."""
    if not REGISTRY_PATH.exists():
        return {"datasets": {}}

    content = REGISTRY_PATH.read_text(encoding="utf-8")
    if yaml is None:
        return _minimal_parse_registry(content)

    return yaml.safe_load(content) or {"datasets": {}}


def get_dataset_path(dataset_id: str, fallback: str | Path | None = None) -> Path:
    """
    Resolve a dataset path from the registry.

    If the registry is missing, the dataset entry is absent, or the entry has no path,
    the provided fallback is used instead. Relative fallbacks are resolved from repo root.
    """

    datasets = load_registry().get("datasets", {})
    entry = datasets.get(dataset_id, {})
    path_value = entry.get("path") if isinstance(entry, dict) else None

    if path_value:
        return (ROOT_DIR / str(path_value)).resolve()

    if fallback is None:
        raise KeyError(f"Dataset '{dataset_id}' not found in {REGISTRY_PATH}")

    fallback_path = Path(fallback)
    if not fallback_path.is_absolute():
        fallback_path = ROOT_DIR / fallback_path
    return fallback_path.resolve()
